### PR TITLE
api-fetch: Type several of the middlewares

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds type checking to the following middleware:
* namespace-endpoint
* preloading
* root-url
* user-locale

It also simplifies the types used in the `nonce` middleware.

It also corrects the return types for the `ApiFetchMiddleware` and its `next` function. These eventually always return a `Promise` as the final middleware eventually calls to the default handler. Each middleware itself can also return a Promise instead of calling `next` (see `preloading` and `media-upload`).

Part of #18838 

## How has this been tested?
Type check passes. No runtime changes except converting some `function`s to arrow functions so that we can type them as `ApiFetchMiddleware` instead of having to annotate the parameter types manually (see the changes to the `nonce` middleware for an example of this).

## Types of changes
Non-breaking changes.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/api-fetch/src/middlewares/nonce.js [hunk 1]
```diff
@@ -4,11 +4,9 @@
  */
 function createNonceMiddleware( nonce ) {
 	/**
-	 * @param {import('../types').ApiFetchRequestProps} options
-	 * @param {(options: import('../types').ApiFetchRequestProps) => import('../types').ApiFetchRequestProps} next
-	 * @return {import('../types').ApiFetchRequestProps} The enhanced request.
+	 * @type {import('../types').ApiFetchMiddleware & { nonce: string }}
```
Comment (tyxla) [inline packages/api-fetch/src/middlewares/nonce.js:7]: I guess we could do a bit better with typing the `{ nonce: string }` part.

**Review by tyxla — APPROVED** [review]
Looks good, thank you 👍

File: packages/api-fetch/src/middlewares/nonce.js (hunk 1 above)
Comment (sarayourfriend) [reply packages/api-fetch/src/middlewares/nonce.js:7]: What would you suggest? This is the only way I know how to type a function with extra properties on it other than using a new interface:

```ts
interface ApiFetchNonceMiddleware extends ApiFetchMiddleware {
    nonce: string;
}
```

but that's effectively the same as what I have inline here.

Comment (tyxla) [reply packages/api-fetch/src/middlewares/nonce.js:7]: I agree, that's not much better 👍  Let's keep as-is for now

**Comment by sarayourfriend** [comment]
Thanks for the review @tyxla!

**Pull request merged**